File: leetcode/binary_tree/level_order.rb
def level_order(root)
  return [] if root.nil?
  curr_q = [root]
  levels = []

  while !curr_q.empty?
    level = []
    next_q = []

    while !curr_q.empty?
      node = curr_q.shift
      level << node.val

      next_q << node.left if node.left
      next_q << node.right if node.right
    end

    levels << level
    curr_q = next_q
  end

  levels
end

File: leetcode/binary_tree/levelOrder.js
var levelOrder = function(root) {
  let parents = [root];
  const results = [];

  while parents.length > 0 {
    let children = [];

    while parents.length > 0 {
      let parent = parents.shift();

      results.push(parent.val);

      if (parent.left) children.push(parent.left);
      if (parent.right) children.push(parent.right);
    }

    parents = children;
  }

  return results;
};
